### ISSUE TITLE
with_transaction rollback scenario crashing if the transaction fun inside the try block crashes.

### ISSUE THREAD

**Issue by silverblaze404** [epgsql/epgsql#295]
Crash - 

`** exception exit: {noproc,{gen_server,call,
                                       [<0.70445.0>,
                                        {command,epgsql_cmd_squery,"ROLLBACK"},
                                        infinity]}}
     in function  gen_server:call/3 (gen_server.erl, line 401)
     in call from epgsql:with_transaction/3`

Code block -

`with_transaction(C, F, Opts0) ->
    Opts = to_map(Opts0),
    Begin = case Opts of
                #{begin_opts := BeginOpts} ->
                    [<<"BEGIN ">> | BeginOpts];
                _ -> <<"BEGIN">>
            end,
    try
        {ok, [], []} = squery(C, Begin),
        R = F(C),
        {ok, [], []} = squery(C, <<"COMMIT">>),
        case Opts of
            #{ensure_committed := true} ->
                {ok, CmdStatus} = get_cmd_status(C),
                (commit == CmdStatus) orelse error({ensure_committed_failed, CmdStatus});
            _ -> ok
        end,
        R
    catch
        ?WITH_STACKTRACE(Type, Reason, Stack)
            squery(C, "ROLLBACK"),
            case maps:get(reraise, Opts, true) of
                true ->
                    erlang:raise(Type, Reason, Stack);
                false ->
                    {rollback, Reason}
            end
    end.`

Here ROLLBACK squery will always crash because the C (Pid) crashes within the try block.

**Comment by seriyps** [comment]
For some reason the epgsql connection process have crashed. Can you show what fun do you pass to `with_transaction`?

**Comment by silverblaze404** [comment]
Actually i am using `epgsql:equery(Conn, Sql, Parameters)` here.

What I am doing is passing integer values for columns greater than the int4 (which is the column's data type). This causes the equery to throw the integer overflow error.

**Comment by seriyps** [comment]
Yes, we have a hard range checks there https://github.com/epgsql/epgsql/blob/7ba52768cf0ea7d084df24d4275a88eef4db13c2/src/datatypes/epgsql_codec_integer.erl#L31

It will crash even without `with_transaction`. My only suggestion here will be to do the overflow check before you pass it to `epgsql`.

We could return some error report instead of crashing as well, but it would require some noticeable amount of work and I'm not sure it is that beneficial.

**Comment by silverblaze404** [comment]
Yeah the rollback might not be beneficial to implement the rollback scenario, but I think it should raise the original stacktrace  so that the caller knows what caused the transaction to fail.

**Comment by eproxus** [comment]
We have the same problem in production with this library. When certain things goes wrong in the transaction, `epgsql` tries to do a rollback on a crashed connection process, which hides the original error. I think the call to `squery` that tries to do the ROLLBACK should be wrapped in another `try ... catch` block to handle if the process is missing (no rollback actually needed in that case) and return the original error instead of implicitly hiding it.

**Comment by seriyps** [comment]
If `squery(C, "ROLLBACK")` crashes with `noproc`, it means there should be `gen_server` crash report or supervisor report somewhere in the logs... I'm still thinking just doing `catch squery(C, "ROLLBACK")` is a good idea. Because this way we may swallow some important error message that rollback would generate...

Sidenote: out of all the API functions in `epgsql` the one that created most discussions and support requests is `with_transaction`, which doesn't really belong to the "postgresql driver/connector", because it is more high level than the others. So maybe we should just deprecate and remove it eventually? Put its code to README as an example of how it can be impleented.. So users can implement high-level constructions in a way that better fits their architecture?

**Comment by eproxus** [comment]
@seriyps Yeah, we do see those `gen_server` crashes all over our error reporting system, but they cannot be acted on since there's no connection to the actual process / stack trace that lead to the error so they are practically useless.

Regardless of if `with_transaction` exists in `epgsql` or not, the equivalent code would just produce the same problem, no? The issue seems to be that certain "valid" queries (i.e. somewhat valid looking SQL, that Postgres can still fail to execute for some reason) crashes the connection process and in turn makes the rollback attempt fail in the "client" process. However, these two errors are completely disconnected in the architecture of `epgsql`. I think the root cause lies in the design of this interaction rather than what exactly the client does with the result.

In the end, I think the behavior is not very "functional" in that errors have nasty and complicated side effects inside and outside the library.
